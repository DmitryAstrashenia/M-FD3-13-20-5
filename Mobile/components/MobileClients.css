.MobileClients{
    width: 800px;
}

.MobileClients .edit{
    width: 100%;
    text-align: center;
    padding: 0;
    font-style: italic;
}

.MobileClients table td{
    width: 100px;
}

.MobileClients table,th,td{
    border: solid 1px black;
    text-align: center;
}

.MobileClients table{
    width: 100%;
}

.MobileClients .backgroundColorGreen{
    background-color: green;
}

.MobileClients .backgroundColorRed{
    background-color: red;
}