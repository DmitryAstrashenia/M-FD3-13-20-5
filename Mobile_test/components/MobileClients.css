.mobileClients{
    width: 800px;
}

.mobileClients .edit{
    width: 100%;
    text-align: center;
    padding: 0;
    font-style: italic;
}

.mobileClients table td{
    width: 100px;
}

.mobileClients table,th,td{
    border: solid 1px black;
    text-align: center;
}

.mobileClients table{
    width: 100%;
}

.MobileClients .backgroundColorGreen{
    background-color: green;
}

.MobileClients .backgroundColorRed{
    background-color: red;
}