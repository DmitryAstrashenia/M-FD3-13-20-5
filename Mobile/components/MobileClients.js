"use strict";
import React from "react";
import "./MobileClients";
import { myEvents } from "./events";
import Client from "./Client";
import PropTypes from "prop-types";
import deepEqual from "deep-equal";

class MobileClients extends React.Component {
  static propTypes = {
    clients: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    ),
  };

  state = {
    clients: this.props.clients,
    statusActiveClicked: null,
    statusBlockedClicked: null,
    edit: null,
  };

  allClients = this.props.clients;

  answerDeleteClicked = (id) => {
    let newClients = this.allClients.filter((e) => e.id != id);
    this.setState({
      clients: newClients,
      edit: null,
    });
    this.allClients = newClients;
    if (this.state.statusActiveClicked) {
      this.answerActiveClicked();
    }
    if (this.state.statusBlockedClicked) {
      this.answerBlockedClicked();
    }
  };

  answerAllClicked = () => {
    let isEqual = deepEqual(this.allClients, this.state.clients);
    if (!isEqual) {
      this.setState({
        clients: this.allClients,
        statusActiveClicked: 0,
        statusBlockedClicked: 0,
        edit: null,
      });
    }
  };

  answerActiveClicked = () => {
    let newClients = this.allClients.filter((client) => client.balance >= 0);
    let isEqual = deepEqual(newClients, this.state.clients);
    if (!isEqual) {
      this.setState({
        clients: newClients,
        statusActiveClicked: 1,
        statusBlockedClicked: 0,
        edit: null,
      });
    }
  };

  answerBlockedClicked = () => {
    let newClients = this.allClients.filter((client) => client.balance <= 0);
    let isEqual = deepEqual(newClients, this.state.clients);
    if (!isEqual) {
      this.setState({
        clients: newClients,
        statusBlockedClicked: 1,
        statusActiveClicked: 0,
        edit: null,
      });
    }
  };

  answerAddClicked = () => {
    let maxId = 0;
    this.allClients.forEach((e) => {
      if (e.id > maxId) {
        maxId = e.id;
      }
    });
    let newClients = [...this.allClients];
    let newClient = {
      id: 5 + maxId,
      fam: null,
      im: null,
      otch: null,
      balance: this.state.statusBlockedClicked ? -1 : 1,
    };
    newClients.push(newClient);

    this.setState({
      clients: newClients,
      edit: 1,
    });
    this.allClients = newClients;
    if (this.state.statusActiveClicked) {
      this.answerActiveClicked();
    }
    if (this.state.statusBlockedClicked) {
      this.answerBlockedClicked();
    }
  };

  answerSaveClicked = (id, fam, im, otch, balance) => {
    let newClients = [...this.allClients];
    newClients.forEach((client, i) => {
      if (client.id == id) {
        let newClient = { ...client };
        newClient.fam = fam;
        newClient.im = im;
        newClient.otch = otch;
        newClient.balance = Number(balance);
        newClients[i] = newClient;
      }
    });
    this.setState({
      clients: newClients,
    });
    this.allClients = newClients;
    if (this.state.statusActiveClicked) {
      this.answerActiveClicked();
    }
    if (this.state.statusBlockedClicked) {
      this.answerBlockedClicked();
    }
  };

  componentDidMount = () => {
    myEvents.addListener("answerDeleteClicked", this.answerDeleteClicked);
    myEvents.addListener("answerSaveClicked", this.answerSaveClicked);
  };

  componentWillUnmount = () => {
    myEvents.removeListener("answerDeleteClicked", this.answerDeleteClicked);
    myEvents.removeListener("answerSaveClicked", this.answerSaveClicked);
  };

  render() {
    console.log("renderMobileClients");

    let clientsCode = this.state.clients.map((client) => (
      <Client key={client.id} client={client} edit={this.state.edit} />
    ));
    return (
      <div className={"MobileClients"}>
        <input type="button" value="Все" onClick={this.answerAllClicked} />
        <input
          type="button"
          value="Активные"
          onClick={this.answerActiveClicked}
        />
        <input
          type="button"
          value="Заблокированные"
          onClick={this.answerBlockedClicked}
        />
        <hr />
        <hr />
        <table>
          <thead>
            <tr>
              <th>Фамилия</th>
              <th>Имя</th>
              <th>Отчество</th>
              <th>Баланс</th>
              <th>Статус</th>
              <th>Редактировать</th>
              <th>Удалить</th>
            </tr>
          </thead>
          <tbody>{clientsCode}</tbody>
        </table>
        <hr />
        <input
          type="button"
          value="Добавить клиента"
          onClick={this.answerAddClicked}
        />
      </div>
    );
  }
}

export default MobileClients;
