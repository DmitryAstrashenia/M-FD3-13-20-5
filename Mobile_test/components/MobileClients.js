"use strict";
import React from "react";
import "./MobileClients.css";
import { myEvents } from "./events";
import Client from "./Client";
import PropTypes from "prop-types";
import { editArr } from "./module/editArr";

class MobileClients extends React.PureComponent {
  static propTypes = {
    clients: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
      })
    ),
  };

  state = {
    clients: this.props.clients,
    statusActive: null,
    statusBlocked: null,
    edit: null,
  };

  filterArr = (id, arr) => {
    let newClients = arr.filter((e) => e.id != id);
    return newClients;
  };

  deleteUser = (id) => {
    this.setState({
      clients: this.filterArr(id, this.state.clients),
    });
  };

  btnViewAllUser = () => {
    if (!this.state.statusActive != null || !this.state.statusBlocked != null) {
      this.setState({
        statusActive: null,
        statusBlocked: null,
        edit: null,
      });
    }
  };

  btnViewActiveUser = () => {
    if (this.state.statusActive != 1) {
      this.setState({
        statusActive: 1,
        statusBlocked: null,
        edit: null,
      });
    }
  };

  btnViewBlockedUser = () => {
    if (this.state.statusBlocked != 1) {
      this.setState({
        statusActive: null,
        statusBlocked: 1,
        edit: null,
      });
    }
  };

  btnAddUser = () => {
    let maxId = 0;
    this.state.clients.forEach((e) => {
      if (e.id > maxId) {
        maxId = e.id;
      }
    });
    let newClients = [...this.state.clients];
    let newClient = {
      id: 5 + maxId,
      fam: String(""),
      im: String(""),
      otch: String(""),
      balance: this.state.statusBlocked ? -1 : 1,
    };
    newClients.push(newClient);

    this.setState({
      clients: newClients,
      edit: true,
    });
  };

  saveClicked = (newClient) => {
    let newClients = [...this.state.clients];
    this.setState({
      clients: editArr(newClients, newClient),
      edit: null,
    });
  };

  componentDidMount = () => {
    myEvents.addListener("answerDeleteClicked", this.deleteUser);
    myEvents.addListener("answerSaveClickedFromEditClient", this.saveClicked);
  };

  componentWillUnmount = () => {
    myEvents.removeListener("answerDeleteClicked", this.deleteUser);
    myEvents.removeListener(
      "answerSaveClickedFromEditClient",
      this.saveClicked
    );
  };

  render() {
    console.log("renderMobileClients");
    let newClients = [];
    let clientsCode = [];
    if (this.state.statusActive) {
      newClients = this.state.clients.filter((client) => client.balance >= 0);
      clientsCode = newClients.map((client) => (
        <Client key={client.id} client={client} edit={this.state.edit} />
      ));
    } else if (this.state.statusBlocked) {
      newClients = this.state.clients.filter((client) => client.balance < 0);
      clientsCode = newClients.map((client) => (
        <Client key={client.id} client={client} edit={this.state.edit} />
      ));
    } else {
      clientsCode = this.state.clients.map((client) => (
        <Client key={client.id} client={client} edit={this.state.edit} />
      ));
    }
    return (
      <div className={"mobileClients"}>
        <input
          type="button"
          id="btnViewAllUser"
          value="Все"
          onClick={this.btnViewAllUser}
        />
        <input
          type="button"
          id="btnViewActiveUser"
          value="Активные"
          onClick={this.btnViewActiveUser}
        />
        <input
          type="button"
          id="btnViewBlockedUser"
          value="Заблокированные"
          onClick={this.btnViewBlockedUser}
        />
        <hr />
        <hr />
        <table>
          <thead>
            <tr>
              <th>Фамилия</th>
              <th>Имя</th>
              <th>Отчество</th>
              <th>Баланс</th>
              <th>Статус</th>
              <th>Редактировать</th>
              <th>Удалить</th>
            </tr>
          </thead>
          <tbody>{clientsCode}</tbody>
        </table>
        <hr />
        <input
          type="button"
          id={"answerAddClicked"}
          value="Добавить клиента"
          onClick={this.btnAddUser}
        />
      </div>
    );
  }
}

export default MobileClients;
