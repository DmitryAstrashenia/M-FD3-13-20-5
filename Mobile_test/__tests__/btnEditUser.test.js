"use strict";

import React from "react";
import renderer from "react-test-renderer";
import MobileClients from "../components/MobileClients";

let clientsArr = [
  { id: 101, fam: "Иванов", im: "Иван", otch: "Иванович", balance: 200 },
  { id: 105, fam: "Сидоров", im: "Сидор", otch: "Сидорович", balance: 250 },
  { id: 110, fam: "Петров", im: "Пётр", otch: "Петрович", balance: 180 },
  {
    id: 120,
    fam: "Григорьев",
    im: "Григорий",
    otch: "Григорьевич",
    balance: -220,
  },
];

test("работа MobileClients.answerEditClicked", () => {
  const component = renderer.create(
    <MobileClients
      clients={clientsArr}
      statusActiveClicked={null}
      statusBlockedClicked={null}
      edit={null}
    />
  );

  let componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();

  let buttonElemEdit = component.root.find(
    (el) => el.props.id == Number(101) + "_edit"
  );

  buttonElemEdit.props.onClick();

  componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();

  buttonElemEdit = component.root.find(
    (el) => el.props.id == Number(101) + "_save"
  );

  buttonElemEdit.props.onClick();

  componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();
});
