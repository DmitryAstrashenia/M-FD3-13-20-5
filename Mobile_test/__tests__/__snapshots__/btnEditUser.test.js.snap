// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа MobileClients.answerEditClicked 1`] = `
<div
  className="mobileClients"
>
  <input
    id="btnViewAllUser"
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    id="btnViewActiveUser"
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    id="btnViewBlockedUser"
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <hr />
  <hr />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="backgroundColorGreen"
        >
          active
        </td>
        <td>
          <input
            id="101_edit"
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            id="101_delete"
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="backgroundColorGreen"
        >
          active
        </td>
        <td>
          <input
            id="105_edit"
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            id="105_delete"
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Пётр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="backgroundColorGreen"
        >
          active
        </td>
        <td>
          <input
            id="110_edit"
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            id="110_delete"
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="backgroundColorRed"
        >
          blocked
        </td>
        <td>
          <input
            id="120_edit"
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            id="120_delete"
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <hr />
  <input
    id="answerAddClicked"
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;

exports[`работа MobileClients.answerEditClicked 2`] = `
<div
  className="mobileClients"
>
  <input
    id="btnViewAllUser"
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    id="btnViewActiveUser"
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    id="btnViewBlockedUser"
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <hr />
  <hr />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <input
            className="edit"
            defaultValue="Иванов"
            type="text"
          />
        </td>
        <td>
          <input
            className="edit"
            defaultValue="Иван"
            type="text"
          />
        </td>
        <td>
          <input
            className="edit"
            defaultValue="Иванович"
            type="text"
          />
        </td>
        <td>
          <input
            className="edit"
            defaultValue={200}
            type="text"
          />
        </td>
        <td
          className="backgroundColorGreen"
        >
          active
        </td>
        <td>
          <input
            id="101_save"
            onClick={[Function]}
            type="button"
            value="Сохранить"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="backgroundColorGreen"
        >
          active
        </td>
        <td>
          <input
            id="105_edit"
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            id="105_delete"
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Пётр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="backgroundColorGreen"
        >
          active
        </td>
        <td>
          <input
            id="110_edit"
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            id="110_delete"
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="backgroundColorRed"
        >
          blocked
        </td>
        <td>
          <input
            id="120_edit"
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            id="120_delete"
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <hr />
  <input
    id="answerAddClicked"
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;

exports[`работа MobileClients.answerEditClicked 3`] = `
<div
  className="mobileClients"
>
  <input
    id="btnViewAllUser"
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    id="btnViewActiveUser"
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    id="btnViewBlockedUser"
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <hr />
  <hr />
  <table>
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <input
            className="edit"
            defaultValue="Иванов"
            type="text"
          />
        </td>
        <td>
          <input
            className="edit"
            defaultValue="Иван"
            type="text"
          />
        </td>
        <td>
          <input
            className="edit"
            defaultValue="Иванович"
            type="text"
          />
        </td>
        <td>
          <input
            className="edit"
            defaultValue={200}
            type="text"
          />
        </td>
        <td
          className="backgroundColorGreen"
        >
          active
        </td>
        <td>
          <input
            id="101_save"
            onClick={[Function]}
            type="button"
            value="Сохранить"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="backgroundColorGreen"
        >
          active
        </td>
        <td>
          <input
            id="105_edit"
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            id="105_delete"
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Пётр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="backgroundColorGreen"
        >
          active
        </td>
        <td>
          <input
            id="110_edit"
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            id="110_delete"
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="backgroundColorRed"
        >
          blocked
        </td>
        <td>
          <input
            id="120_edit"
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            id="120_delete"
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <hr />
  <input
    id="answerAddClicked"
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;
